Tensor.contiguous, Tensor.is_contiguous batch rule

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47632 Batching rule for torch.squeeze(tensor)
* #47628 Fix transpose and sum batching rule
* #47625 Make kwargs argument optional in _batched_grad_test
* **#47621 Tensor.contiguous, Tensor.is_contiguous batch rule**

Followup to #47365.

is_contiguous on BatchedTensorImpl is implemented as:
- Whenever one creates a BatchedTensorImpl, we cache the strides of the
per-examples, just like how we cache the sizes of the per-examples.
- With the cached strides, we use TensorImpl::refresh_contiguous() to
compute if the tensor is contiguous or not.
- is_contiguous checks the `is_contiguous_` flag that
refresh_contiguous() populates.

Both contiguous and is_contiguous only support torch.contiguous_format.
I'm not sure what the semantics should be for other memory formats; they
are also rank dependent (e.g., channels_last tensor must have 4
dimensions) which makes this a bit tricky.

Test Plan:
- new tests

Differential Revision: [D24840975](https://our.internmc.facebook.com/intern/diff/D24840975)

zou3519 committed:
Tensor.contiguous, Tensor.is_contiguous batch rule

Followup to #47365.

is_contiguous on BatchedTensorImpl is implemented as:
- Whenever one creates a BatchedTensorImpl, we cache the strides of the
per-examples, just like how we cache the sizes of the per-examples.
- With the cached strides, we use TensorImpl::refresh_contiguous() to
compute if the tensor is contiguous or not.
- is_contiguous checks the `is_contiguous_` flag that
refresh_contiguous() populates.

Both contiguous and is_contiguous only support torch.contiguous_format.
I'm not sure what the semantics should be for other memory formats; they
are also rank dependent (e.g., channels_last tensor must have 4
dimensions) which makes this a bit tricky.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Tensor.contiguous, Tensor.is_contiguous batch rule"


Followup to #47365.

is_contiguous on BatchedTensorImpl is implemented as:
- Whenever one creates a BatchedTensorImpl, we cache the strides of the
per-examples, just like how we cache the sizes of the per-examples.
- With the cached strides, we use TensorImpl::refresh_contiguous() to
compute if the tensor is contiguous or not.
- is_contiguous checks the `is_contiguous_` flag that
refresh_contiguous() populates.

Both contiguous and is_contiguous only support torch.contiguous_format.
I'm not sure what the semantics should be for other memory formats; they
are also rank dependent (e.g., channels_last tensor must have 4
dimensions) which makes this a bit tricky.

Test Plan:
- new tests

Differential Revision: [D24840975](https://our.internmc.facebook.com/intern/diff/D24840975)

[ghstack-poisoned]

albanD commented on test/test_vmap.py:
```diff
@@ -1323,6 +1343,55 @@ def foo(x):
+        contig = torch.randn(B0, 2, 7)
+        self.assertEqual(vmap(foo)(contig), torch.ones(B0))
+
+        noncontig = torch.randn(2, B0, 7)
```
nit: could you add a test here to make sure that this is properly returning false even if you `permute(1, 0, 2)`?

albanD commented on aten/src/ATen/BatchedTensorImpl.cpp:
```diff
@@ -19,13 +19,18 @@ BatchedTensorImpl::BatchedTensorImpl(Tensor value, BatchDims bdims)
+    strides_.push_back(value_strides.at(actual_dim));
   }
   refresh_numel();
+  refresh_contiguous();
```
Does this function need extra care?
What if I do `(B1, B0, 2, 3)` and permute to `(B0, B1, 2, 3)` it should still be contiguous even though the physical Tensor is not?

zou3519 replied:
refresh_contiguous() uses the `sizes_` and `strides_` field of the TensorImpl to determine if things are contiguous. Because the sizes_ and strides_ fields of BatchedTensorImpl are the per-example sizes and strides, the "is contiguous" computation is independent of the batch dims. 

> What if I do (B1, B0, 2, 3) and permute to (B0, B1, 2, 3) it should still be contiguous even though the physical Tensor is not?

Yes, it should still be contiguous. In this example, `sizes_ = [2, 3]` and `strides_ = [3, 1]`. `refresh_contiguous()` operates on those and those check out to be contiguous. I'll add an extra test case for this.

zou3519 replied:
For completeness:
- code for [refresh_contiguous](https://github.com/pytorch/pytorch/blob/f2eac5df1840b65b24ef2f1c3bd6332f654935b6/c10/core/TensorImpl.h#L1538-L1539) (we only really care about the first line that calls compute_contiguous)
- code for [compute_contiguous](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.cpp#L94-L110)

albanD replied:
Sounds good! Thanks for the links!

zou3519 committed:
Update on "Tensor.contiguous, Tensor.is_contiguous batch rule"


Followup to #47365.

is_contiguous on BatchedTensorImpl is implemented as:
- Whenever one creates a BatchedTensorImpl, we cache the strides of the
per-examples, just like how we cache the sizes of the per-examples.
- With the cached strides, we use TensorImpl::refresh_contiguous() to
compute if the tensor is contiguous or not.
- is_contiguous checks the `is_contiguous_` flag that
refresh_contiguous() populates.

Both contiguous and is_contiguous only support torch.contiguous_format.
I'm not sure what the semantics should be for other memory formats; they
are also rank dependent (e.g., channels_last tensor must have 4
dimensions) which makes this a bit tricky.

Test Plan:
- new tests

Differential Revision: [D24840975](https://our.internmc.facebook.com/intern/diff/D24840975)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

ezyang commented:
@zou3519 think this PR should get someone from the memory layout crowd to take a look? (I tentatively added Vitaly but maybe there is a better person; Vitaly would you know?)

VitalyFedyunin commented:
I will take a look today

zou3519 commented:
@VitalyFedyunin please let me know if you have any questions and/or if I can provide more context!

Some context
--------------
The context behind this change is that we'd like to define semantics for calling .contiguous() inside vmap. For example, let Bi be dimensions that are being vmapped over.
```
x = torch.randn(B0, B1, 3, 5)
y = vmap(vmap(Tensor.contiguous))(x)
```

For memory format = torch.contiguous_format, we've defined the semantics of "is_contiguous" as "check that each per-example inside of vmap is contiguous". So, in terms of this example, because each slice of x of shape [3, 5] is contiguous, the contiguous() operator will return a `y` that is the same as `x` (`y is x`).

This gets a bit more interesting when `x` is not contiguous, but the per-examples are contiguous. 
```
x = torch.randn(B1, B0, 3, 5).transpose(0, 1) # x is not contiguous
assert not x.is_contiguous()
y = vmap(vmap(Tensor.contiguous))(x)
```
Here, since the per-examples are contiguous, `assert y is x` is True.

What this PR does
------------------
This PR allows users to query `is_contiguous` and call `.contiguous()` inside of vmap, but only if the passed in memory-format is `torch.contiguous_format`. I didn't want to think about the semantics of other memory formats because that seems complicated.

For example, if someone does a vmap over an NHWC contiguous tensor and calls F.conv2d, the per-samples don't actually "have memory format torch.channels_last" because the per-samples are 3D tensors and channels_last memory format requires 4D tensors

zou3519 committed:
Update on "Tensor.contiguous, Tensor.is_contiguous batch rule"


Followup to #47365.

is_contiguous on BatchedTensorImpl is implemented as:
- Whenever one creates a BatchedTensorImpl, we cache the strides of the
per-examples, just like how we cache the sizes of the per-examples.
- With the cached strides, we use TensorImpl::refresh_contiguous() to
compute if the tensor is contiguous or not.
- is_contiguous checks the `is_contiguous_` flag that
refresh_contiguous() populates.

Both contiguous and is_contiguous only support torch.contiguous_format.
I'm not sure what the semantics should be for other memory formats; they
are also rank dependent (e.g., channels_last tensor must have 4
dimensions) which makes this a bit tricky.

Test Plan:
- new tests

Differential Revision: [D24840975](https://our.internmc.facebook.com/intern/diff/D24840975)

[ghstack-poisoned]

VitalyFedyunin reviewed (approved):
Looks good to be, but might require special attention later if we decide to introduce non-batched 3d (HWC) channel last tensors later.

zou3519 committed:
Update on "Tensor.contiguous, Tensor.is_contiguous batch rule"


Followup to #47365.

is_contiguous on BatchedTensorImpl is implemented as:
- Whenever one creates a BatchedTensorImpl, we cache the strides of the
per-examples, just like how we cache the sizes of the per-examples.
- With the cached strides, we use TensorImpl::refresh_contiguous() to
compute if the tensor is contiguous or not.
- is_contiguous checks the `is_contiguous_` flag that
refresh_contiguous() populates.

Both contiguous and is_contiguous only support torch.contiguous_format.
I'm not sure what the semantics should be for other memory formats; they
are also rank dependent (e.g., channels_last tensor must have 4
dimensions) which makes this a bit tricky.

Test Plan:
- new tests

Differential Revision: [D24840975](https://our.internmc.facebook.com/intern/diff/D24840975)

[ghstack-poisoned]
